Add support for iOS UndoManager

fbcouch opened the pull request:
This PR is modifies the framework to allow the Undo/Redo support added by `_TextEditingHistory` to interact with the iOS `UndoManager`. That should enable 3-finger/shake gestures to undo/redo text editing and enable undo/redo from the iPad keyboard.

I don't love this approach, since it feels a little bit hacky and requires a breaking change to the TextInputClient interface. It's possible #98280 could enable a better way of doing this.

~~I haven't had a chance to test this out – still waiting for the engine to compile, but wanted to plant the flag on this and possibly feedback on the approach.~~

Related to #34749 and #77614

Engine PR: https://github.com/flutter/engine/pull/31415

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

fbcouch commented:
I believe this is a reasonable approach – I think it may make sense to swap over to something along the lines of `setUndoManagerState`, though, similar to `updateEditingState`, where the engine can just set up the undoManager as needed, rather than having the framework manage it through three separate calls.

fbcouch commented:
@justinmc I'd like to get your thoughts on the framework side of things – right now I'm adding a method to `TextInputClient`, which will make this a breaking change. I'm wondering if it would be better to add a property to the `TextInputConnection` for some kind of new `UndoDelegate`, and have the `TextEditingHistory` widget set itself as the current undo delegate on the connection when it receives focus. That way we could avoid that breaking change. Does that make sense?

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2959,6 +2959,23 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+  void handleKeyboardUndo(String direction) {
+    if (_editableKey.currentContext == null)
+        return;
+    if (direction == 'undo') {
```
Nit: Would it be better to send a boolean from the engine?  It would save a little bit of space at least.

If not, maybe parse the string into an enum or compare it to constants.

justinmc reviewed (commented):
Just to throw out a crazy idea: What if everything in the engine through to the platform message channel were separate from text editing?  Since NSUndoManager itself is not specific to text editing, hypothetically this could be generic to any kind of undo/redo.  In EditableText you would specifically use it for text editing.  Just a thought I had, not necessarily a good idea.

Otherwise I like the idea you mentioned of an "UndoDelegate".  Kind of similar to [DeltaTextInputClient](https://github.com/flutter/flutter/blob/660c15ef7f6b70b9a8d0103b17ea6d157f6e0418/packages/flutter/lib/src/services/text_input.dart#L1191)?

chunhtai commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2959,6 +2959,23 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+    if (direction == 'undo') {
+      const UndoTextIntent intent = UndoTextIntent(SelectionChangedCause.keyboard);
+      final Action<UndoTextIntent>? action = Actions.maybeFind<UndoTextIntent>(_editableKey.currentContext!, intent: intent);
+      if (action != null)
```
this can be moved out of the both if(direction == 'undo') and else case so that we don't need to duplicate code

fbcouch commented:
Thinking about this more, I really like the idea of separating this out from the text editing details. That way a drawing app could make use of it. I haven't had a chance to work on it this week (in the middle of moving), but I'm hoping to be back at it next week.

Does it make sense to still use a FocusNode to know which undo client should receive messages or should that be something left up to the implementers? I think FocusNodes make sense for text editing, but I'm not sure about drawing. I think it still could, but maybe best left to others to decide.

My initial thought is it would make sense to have some kind of `UndoActions` widget (not sure on the name) that takes an `UndoController` or some such that provides methods like `canUndo`, `canRedo`, `undo()` and `redo()`. That's where the question about focus nodes comes in...should `UndoActions` take a focusNode in order to automatically manage focus, or would it make more sense to have a method on `UndoController` to become the receiver of undo messages?

fbcouch commented:
I converted this back to draft status for now, since this is a work in progress.

I added a new widget, `UndoHistory`, and a new service, `UndoManager` to handle the separate communication and make this more generic.

I haven't updated `EditableText` to use it yet, but here's an example app that I used to mock up an Undo/Redo UI around a `TextField`:

<details>
<summary>lib/main.dart</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextEditingController _controller = TextEditingController();
  final FocusNode _focusNode = FocusNode();
  final GlobalKey<UndoHistoryState> _undoKey = GlobalKey<UndoHistoryState>();

  TextStyle? get enabledStyle => Theme.of(context).textTheme.bodyMedium;
  TextStyle? get disabledStyle => Theme.of(context).textTheme.bodyMedium?.copyWith(color: Colors.grey);

  bool _canUndo = false, _canRedo = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            UndoHistory<TextEditingValue>(
              key: _undoKey,
              value: _controller,
              onTriggered: (value) {
                _controller.value = _controller.value.copyWith(text: value.text, selection: value.selection);
              },
              focusNode: _focusNode,
              shouldChangeUndoStack: (TextEditingValue? oldValue, TextEditingValue newValue) {
                if (oldValue == null) {
                  return true;
                }

                if (newValue == TextEditingValue.empty) {
                  return false;
                }

                return oldValue.text != newValue.text;
              },
              onUndoStackChanged: () => setState(() {
                _canUndo = _undoKey.currentState?.canUndo ?? false;
                _canRedo = _undoKey.currentState?.canRedo ?? false;
              }),
              child: TextField(
                maxLines: 4,
                controller: _controller,
                focusNode: _focusNode,
              ),
            ),
            Row(
              children: [
                TextButton(
                    child: Text('Undo', style: _canUndo ? enabledStyle : disabledStyle),
                    onPressed: () {
                      _undoKey.currentState?.undo();
                    }),
                TextButton(
                    child: Text('Redo', style: _canRedo ? enabledStyle : disabledStyle),
                    onPressed: () {
                      _undoKey.currentState?.redo();
                    }),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
```


</details>

justinmc commented:
I like that approach, looks really reusable.  Would that replace _TextEditingHistory or is it separate?

fbcouch commented:
Cool, yes, I actually was able to get to the point of replacing `_TextEditingHistory` on Friday: https://github.com/flutter/flutter/pull/98294/files#diff-f5f93c879cef9a102adbf148583de9f3d9a05f9678fdcad7bde106f62200bb34R3193

There are a couple of things I don't love about the implementation right now, so let me know if you have thoughts on them:

- I put the API to access the undo state on `UndoHistoryState`, but it's not very elegant to access from a `TextField(key: _key)` right now:  `(_key.currentState as TextSelectionGestureDetectorBuilderDelegate).editableTextKey.currentState?.undoHistory?.redo()`. I'm wondering if providing more direct access and/or adding another interface on to `_TextFieldState` might make that simpler. `UndoHistoryDelegate` or something like that.
- To provide software undo/redo buttons, I have to manually call `setState` after changes, so that might be something that either the `UndoHistoryDelegate` could provide a stream for, or there could be a stream at another level – maybe even on the `UndoManager` itself.

fbcouch commented:
All right, I swapped this over to use a controller and I think it's a lot cleaner: https://github.com/flutter/flutter/pull/98294/files#diff-0dd319c779a725fd887d8a664e656ce5cf92b62934a821cf2f3815750c630245R56

Let me know how that looks!

justinmc commented:
CC @LongCatIsLooong who is probably also interested in this.  We talked about the 3 finger swipe gesture, etc. before.

justinmc reviewed (commented):
A few quick comments here, I owe you a full review by Monday.  Really excited for this PR.

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -4299,256 +4323,3 @@ class _CopySelectionAction extends ContextAction<CopySelectionTextIntent> {
-
-/// A void function that takes a [TextEditingValue].
-@visibleForTesting
-typedef TextEditingValueCallback = void Function(TextEditingValue value);
```
Is that okay? Should we leave the typedef for now, just in case?

justinmc replied:
We should remove it if it's not used by the framework.

If it passes the customer tests and the Google tests then I think it's ok to try merging it.  If it does break anything, then the migration process should be pretty simple (users should include their own typedef in their app I guess?).

fbcouch commented on packages/flutter/lib/src/services/undo_manager.dart:
```diff
@@ -0,0 +1,106 @@
+  /// Receive undo and redo events from the system's UndoManager.
+  ///
+  /// Setting the [client] will cause [UndoManagerClient.handleKeyboardUndo]
+  /// to be called when the system undo or redo buttons are tapped.
```
No, that's a good point. I updated the comment to include the gestures.

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1261,6 +1261,13 @@ class TextInputConnection {
 
+  /// Requests that the text input plugin set the undo or redo state.
+  ///
+  /// This is currently only used on iOS 9+ to integrate with the native NSUndoManager.
```
This actually needs to be removed since the `TextInput` channel is no longer used by this feature

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 24.
View them at https://flutter-gold.skia.org/cl/github/98294

justinmc reviewed (commented):
Mostly just small things, but one point I want to make sure everyone agrees on is the addition of the new controller (UndoHistoryController).  I left a comment about that below.

Otherwise I really like this approach and the fact that undo/redo isn't tied to text editing.

Also I think the Skia Gold failure is not a problem, maybe try pushing a merge commit and see if it goes away?

fbcouch commented on packages/flutter/lib/src/widgets/undo_history.dart:
```diff
@@ -0,0 +1,447 @@
+///
+/// * [EditableText], which uses the [UndoHistory] widget and allows
+///   control of the underlying history using an [UndoHistoryController].
+class UndoHistoryController extends ValueNotifier<UndoHistoryValue> {
```
You are right that the `UndoHistoryValue` is more like a state than a value, so it is different from the `TextEditingController` in that way. That value is important for propagating the current undo/redo state back to the UI so that custom undo/redo interfaces can be built. The other two ChangeNotifiers are for sending events to allow programmatic undo/redo.

I don't love the idea of adding it to TextEditingController, or at least not _only_ there, so that this can remain more generic: for example, someone could build a drawing interface with it and undo/redo lines on the drawing.

I could get on board with Intents, since we already use those for the keyboard shortcuts, though I agree that might be more confusing.

justinmc replied:
Reading through this again I think I'm onboard with UndoHistoryController as-is.  It makes a lot of sense in the example you included in the examples directory.

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 25.
View them at https://flutter-gold.skia.org/cl/github/98294

justinmc commented on packages/flutter/lib/src/services/undo_manager.dart:
```diff
@@ -0,0 +1,111 @@
+}
+
+/// An interface to receive events from a native UndoManager.
+abstract class UndoManagerClient {
```
I'm currently investigating whether or not we should use a mixin in cases like this instead of an abstract class (to avoid breaking users that extend this if/when a method is added).  I'll let you know what I figure out and then I owe you a full review.

justinmc replied:
[Here is the design doc](https://docs.google.com/document/d/1OxDsf_eot7TlsRn57z-1_dGbvuqnMxO7QEGIAmOCZtA/edit#).  There's still some debate going on.

justinmc replied:
Update: Making this a mixin and using it via `with` instead of `implements` is typically preferred.  Anytime a new method is added here in the future, giving it an empty implementation will avoid breaking changes.  If breaking changes are desired, a method can be added with no implementation.

fbcouch committed:
Implement UndoManager

justinmc commented on examples/api/lib/widgets/undo_history/undo_history_controller.0.dart:
```diff
@@ -0,0 +1,79 @@
+                        child: Text('Undo', style: value.canUndo ? enabledStyle : disabledStyle),
+                        onPressed: () {
+                          _undoController.undo();
+                        }),
```
Here and below, split the closing `})` onto two separate lines.

justinmc commented on packages/flutter/lib/src/services/undo_manager.dart:
```diff
@@ -0,0 +1,111 @@
+
+/// A low-level interface to the system's undo manager.
+///
+/// To receive events from the system UndoManager, create an
```
Nit: "UndoManager" => "undo manager" to match the previous paragraph?

justinmc commented on packages/flutter/lib/src/services/undo_manager.dart:
```diff
@@ -0,0 +1,111 @@
+    _channel.setMethodCallHandler(_handleUndoManagerInvocation);
+  }
+
+  /// Set the [MethodChannel] used to communicate with the system's text input
+  /// control.
+  ///
+  /// This is only meant for testing within the Flutter SDK. Changing this
+  /// will break the ability to input text. This has no effect if asserts are
+  /// disabled.
```
This sounds like TextInputClient, can you tweak them so they're about undo/redo?

justinmc commented on packages/flutter/lib/src/services/undo_manager.dart:
```diff
@@ -0,0 +1,111 @@
+  /// Will be true if there are past values on the stack.
+  bool get canUndo;
+
+  /// Will be false if there are future values on the stack.
```
Should this "false" be "true" or am I misreading it?

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -1607,6 +1615,9 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
   ScrollController? _internalScrollController;
   ScrollController get _scrollController => widget.scrollController ?? (_internalScrollController ??= ScrollController());
 
+  UndoHistoryController? _undoController;
```
Is this never assigned or did I miss that somewhere?

Should UndoHistory create and manage the UndoHistoryController if it's not provided?  Then we could also get rid of _effectiveUndoController.

fbcouch replied:
It does get assigned at the end of line 1619...but...to your other point...I think you might be right that we could let UndoHistory manage that case, since we aren't doing anything here except making sure there is an UndoHistoryController

justinmc commented on packages/flutter/lib/src/widgets/undo_history.dart:
```diff
@@ -0,0 +1,446 @@
+  UndoHistoryController({UndoHistoryValue? value}) : super(value ?? UndoHistoryValue.empty);
+
+  /// Notifies listeners that [undo] has been called.
+  final ChangeNotifier onUndo = ChangeNotifier();
```
Should these ChangeNotifiers be disposed?

chunhtai replied:
Can it be private? It seems to be a way an API just to wire up the undo/redo to the UndoHistory widget.

fbcouch replied:
Possibly, but I think making it public makes this more useful to consumers, for example, if they are implementing their own version of UndoHistory, they would need to listen to the undo/redo notifications.

justinmc commented on packages/flutter/lib/src/widgets/undo_history.dart:
```diff
@@ -0,0 +1,446 @@
+  }
+
+  @override
+  void handlePlatformUndo(String direction) {
```
In general I've been thinking about moving away from the pattern of passing raw, privately defined strings around from these platform methods.

Is there a way you could convert this direction to an enum or something that is publicly defined?  Then in your tests you could also reference that instead of hardcoded "undo" and "redo" strings that I saw in a few places.

justinmc commented on packages/flutter/test/widgets/editable_text_utils.dart:
```diff
@@ -46,6 +47,68 @@ Offset textOffsetToPosition(WidgetTester tester, int offset) {
   return endpoints[0].point + const Offset(0.0, -2.0);
 }
 
+Future<void> sendKeys(
```
Another nit here: I think that should actually be a doc comment with 3 slashes `///`.

justinmc replied:
By the way, I'm glad sendKeys will be reusable now.  I've wanted it in other test files before.

justinmc reviewed (commented):
@fbcouch Are you still available to get this PR ready to merge?  Sorry for the hiatus while I figured out some platform channel stuff.

fbcouch committed:
Change UndoManagerClient to mixin, add UndoDirection enum, other PR changes

fbcouch committed:
Merge branch 'master' into feature/undo-redo-gestures

fbcouch commented:
> @fbcouch Are you still available to get this PR ready to merge? Sorry for the hiatus while I figured out some platform channel stuff.

Yes! I'd love to get this merged in – I think it's the only thing left in the custom flutter version that we run at the moment 😄

fbcouch committed:
Fix analyze failures

fbcouch committed:
Merge branch 'master' into feature/undo-redo-gestures

justinmc commented on packages/flutter/lib/src/services/undo_manager.dart:
```diff
@@ -0,0 +1,129 @@
+
+  static final UndoManager _instance = UndoManager._();
+
+  /// Receive undo and redo events from the system's UndoManager.
```
Nit: Maybe link UndoManager: `[UndoManager].`

justinmc commented on packages/flutter/lib/src/services/undo_manager.dart:
```diff
@@ -0,0 +1,129 @@
+/// [UndoManagerClient] will receive [UndoManagerClient.handlePlatformUndo]
+/// with an [UndoDirection] representing whether the event is "undo" or "redo".
+///
+/// Currently, only iOS has an UndoManagerPlugin implemented on the engine side.
```
Nit: Maybe add the following.  I'm looking at these docs and noticing that someone that lands on this page might not know what this class can actually be used for at a high level.

"On iOS, this can be used to listen to the keyboard undo/redo buttons and the undo/redo gestures."

fbcouch replied:
Someone else should double check me on this, but it seems to work on macOS without any further changes, I think because we're handling the UndoTextIntent and RedoTextIntent.

![Kapture 2022-11-08 at 11 59 53](https://user-images.githubusercontent.com/3249979/200640818-a9b3453a-1923-4d2d-9726-179e0c56175b.gif)

justinmc commented on packages/flutter/lib/src/widgets/undo_history.dart:
```diff
@@ -0,0 +1,466 @@
+
+  @override
+  int get hashCode => Object.hash(
+        canUndo.hashCode,
+        canRedo.hashCode,
+      );
```
Nit: I could be off but I think this way is most aligned with Flutter's code style:

```dart
  int get hashCode => Object.hash(
    canUndo.hashCode,
    canRedo.hashCode,
  );
```

justinmc reviewed (approved):
LGTM :+1:  with nits.  Thanks for being patient with this PR!

CC @chunhtai for secondary review.

chunhtai commented on packages/flutter/lib/src/widgets/undo_history.dart:
```diff
@@ -0,0 +1,466 @@
+    if (nextValue == null) {
+      return;
+    }
+    if (nextValue == _lastValue) {
```
Will this also update if only selection changes?

fbcouch replied:
The logic of whether to update the undo stack or not is controlled by the implementation of `shouldChangeUndoStack` on `UndoHistory` – for `EditableText`, no, only text changes will push onto the stack: https://github.com/flutter/flutter/pull/98294/files#diff-f5f93c879cef9a102adbf148583de9f3d9a05f9678fdcad7bde106f62200bb34R3722

fbcouch committed:
Merge branch 'main' into feature/undo-redo-gestures

fbcouch committed:
Update to handle composing changes from https://github.com/flutter/flutter/pull/108765

fbcouch commented on packages/flutter/lib/src/widgets/undo_history.dart:
```diff
@@ -0,0 +1,466 @@
+  /// If the state would still be the same before and after the undo/redo, this
+  /// will not be called. For example, receiving a redo when there is nothing
+  /// to redo will not call this method.
+  final void Function(T value) onTriggered;
```
If I understand how InputFormatters work correctly, I think this should not be an issue as they run before the value gets here. For example, I tried the CreditCardFormatter from the pattern_formatter package...if I type `1111` in rapid succession so that all ends up on the stack as one undo item, then press `2`, the value pushed onto the stack is `1111 2` (with the space added already by the formatter). When I hit undo, it correctly goes back to `1111` (no space). Pressing redo correctly changes the value to `1111 2`.

chunhtai replied:
No I meant what if the the undo value gets formatted for some reason.
Take your example , `1111 2` if you hit undo this method will be called with `1111`, what if the text format suddenly decide to format the incoming value `1111` to something else like `3333`.  If i understand the code correctly, this will create another undo history right after the undo. Should we throw in this case?

fbcouch replied:
Oh I see what you are saying...I did a RandomNumberFormatter to replace numbers with random digits...it does cause some issues (really what it messes up mostly is being able to redo). The current implementation seems to also have issues with this, so I don't think we are adding new issues here, at least.

So, do you or @justinmc have thoughts on whether that is more of a client concern or something the `UndoHistory` widget should handle?

Essentially, I think we could either handle it by preventing pushes while `onTriggered` is running or passing an optional parameter to `shouldChangeUndoStack` to let clients decide.

I'm leaning toward the latter, since it seems like this is a fairly specific case to EditableText and InputFormatters.

chunhtai replied:
I think this may be a corner case that is not so common in real life. I vote for just assert the value should stay the same after this method is called. If this does crash some use case, we can figure out how to handle it later.  Also at that point we may have a better picture on how we should support this use case

fbcouch replied:
Oh gotcha – I actually already implemented that with this commit: https://github.com/flutter/flutter/pull/98294/commits/b74a5db9dece2a9e40b511fe859844fd2436bce7

I think it works pretty well, though we could change that to make some kind of assert if `_push()` is called while `_duringTrigger` is true or something along those lines.

chunhtai replied:
I am not a big fan of duringTrigger API, It feel a bit confusing to the user if they don't know about this undo issue. I think we should hardcode the behavior for now.

fbcouch replied:
Gotcha. So...should we add an actual `assert` or just a guard clause to prevent pushing that new value onto the undo stack?

I've been playing with assertions in a couple of cases and run into some issues. If I put the assert in `_update`, I end up with failing tests because the value doesn't get updated to the new value on android when composing changes. If I put the assert in `_push`, I can't seem to properly catch that assertion failure in tests because that is called as the result of a change notifier, and I can't figure out how to catch that properly.

At any rate – here's what I ended up going with for now: https://github.com/flutter/flutter/pull/98294/commits/a26bd352845554ce7702380c228b95cf45e4778b – hard coding that behavior in `_push` and updating the documentation for `onTriggered` to describe the behavior.

chunhtai replied:
If we prevent pushing the new value on to the stack, wouldn't that makes the current value on the top of the stack to be inconsistent with whats in the textfield?

Can we add an assert in line 148 in flutter/lib/src/widgets/undo_history.dart
right after the widget.onTriggered(nextValue); is called to verify the value doesn't change?

fbcouch replied:
So, the issue I run into if I try to do that is that composing acts a little weird – I end up in a situation where the following tests fail:
TextEditingHistory does save composing changes on Android (variant: TargetPlatform.android)
TextEditingHistory saves right up to composing change even when throttled (variant: TargetPlatform.android)

I think the issue is that the TextEditingValue ends up changing in that case even though the actual text part of it doesn't. I tried checking that the value is either the new value or the last value, but that didn't resolve the issue.

chunhtai replied:
Do you have a feature branch for the code that causes the error so that i can test and see what was wrong?

fbcouch replied:
Sure – https://github.com/twinsunllc/flutter/tree/feature/undo-redo-gestures-with-assertion

fbcouch replied:
@chunhtai Have you had a chance to take a look at this? Just want to put it on your radar again after the holidays.

chunhtai replied:
sorry for the delay, I will take a look today

chunhtai replied:
The issue is the editable text does not preserve composing in the onTrigger callback. Is this a intended behavior? also cc @justinmc 

If so, the assert should just check for text and selection and omit composing.

Also it should only check for nextValue, it doesn't make sense to check for previousValue. Unless there is a use case I was missing?

justinmc replied:
Changes to composing text are purposely not included in the undo stack (except on Android because Gboard on Android uses composing for everyday English words).  See https://github.com/flutter/flutter/pull/108765.

fbcouch replied:
It's been a little bit since I looked at this in detail, so I will have to refresh my own memory when I get a chance, but if I recall correctly, the reason I was checking the `lastValue` as well is the composing issue – though it still fails those tests, so I'm not sure that works anyway.

Part of the issue is that this is not specific to text editing, so we don't have access to the underlying elements of a TextEditingValue here, we would need to either create an interface to expose some sort of undo/redo-specific value we want compared or farm it out to the clients who are using the UndoHistory. That was my original thinking with adding the optional parameter to shouldChangeUndoStack, that way the clients get to decide how to compare those values and whether changes during the onTrigger are a problem or not.

justinmc replied:
Is your branch [undo-redo-gestures-with-assertion](https://github.com/twinsunllc/flutter/tree/feature/undo-redo-gestures-with-assertion) still up to date, and do you have example code that causes the bugs mentioned here?

I checked out that branch and saw the failures in editable_text_test.dart.  Your change to drop the composing range in onTriggered [0] might not be correct.  Trying it on a native Android app, the composing region seems to remain when undoing a partial word (or maybe it's just reopening).  What if we kept the composing range in editable_text.dart's onTriggered like it was originally, and then modified the test to expect the composing range to exist?

Or maybe I'm oversimplifying and I need to see some example code for the bugs I'm forgetting about.

Here's the original code that didn't drop the composing range:

https://github.com/flutter/flutter/blob/234452fef20f4b1ca199b4a177ed3222979aa2d7/packages/flutter/lib/src/widgets/editable_text.dart#L4285-L4286

And here is a failing test that confirms that the composing range is gone, but maybe it should be changed to include the original composing range:

https://github.com/flutter/flutter/blob/234452fef20f4b1ca199b4a177ed3222979aa2d7/packages/flutter/test/widgets/editable_text_test.dart#L12735

fbcouch replied:
I will trust your judgement on the right way to treat the composing region, I don't know much about it. Tracking the composing changes seems to work, so I went ahead and made those changes and brought that and the assertion over to this branch.

Here's an example that trips the assertion (I modified one of the formatters from the `pattern_formatter` package as a starting point):
<details>
<summary>Example assertion error</summary>

```
import 'dart:math' as math;

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:pattern_formatter/pattern_formatter.dart';

final DateTime launchTime = DateTime.now();

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  static const String _title = 'Flutter Code Sample';

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: _title,
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextEditingController _controller = TextEditingController();
  final FocusNode _focusNode = FocusNode();
  final UndoHistoryController _undoController = UndoHistoryController();

  final FocusNode _counterFocus = FocusNode();
  final ValueNotifier<int> _counter = ValueNotifier<int>(0);
  final UndoHistoryController _counterUndoController = UndoHistoryController();

  TextStyle? get enabledStyle => Theme.of(context).textTheme.bodyMedium;
  TextStyle? get disabledStyle => Theme.of(context).textTheme.bodyMedium?.copyWith(color: Colors.grey);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(launchTime.toIso8601String()),
            UndoHistory<int>(
              value: _counter,
              onTriggered: (int newValue) {
                _counter.value = newValue;
                setState(() {});
              },
              focusNode: _counterFocus,
              controller: _counterUndoController,
              child: Focus(
                focusNode: _counterFocus,
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Text('Current value: ${_counter.value}'),
                    Text('Has Focus: ${_counterFocus.hasFocus}'),
                    TextButton(
                      child: const Text('Increment'),
                      onPressed: () {
                        _counter.value = _counter.value + 1;
                        setState(() {});
                        _counterFocus.requestFocus();
                      },
                    ),
                  ],
                ),
              ),
            ),
            ValueListenableBuilder<UndoHistoryValue>(
              valueListenable: _counterUndoController,
              builder: (context, value, _) {
                return Row(
                  children: [
                    TextButton(
                        child: Text('Undo', style: value.canUndo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _counterUndoController.undo();
                        }),
                    TextButton(
                        child: Text('Redo', style: value.canRedo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _counterUndoController.redo();
                        }),
                  ],
                );
              },
            ),
            TextField(
              maxLines: 4,
              controller: _controller,
              focusNode: _focusNode,
              undoController: _undoController,
              inputFormatters: [
                RandomNumberFormatter(),
              ],
            ),
            ValueListenableBuilder<UndoHistoryValue>(
              valueListenable: _undoController,
              builder: (context, value, _) {
                return Row(
                  children: [
                    TextButton(
                        child: Text('Undo', style: value.canUndo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _undoController.undo();
                        }),
                    TextButton(
                        child: Text('Redo', style: value.canRedo ? enabledStyle : disabledStyle),
                        onPressed: () {
                          _undoController.redo();
                        }),
                  ],
                );
              },
            ),
          ],
        ),
      ),
    );
  }
}

class RandomNumberFormatter extends NumberInputFormatter {
  static final RegExp _digitOnlyRegex = RegExp(r'\d+');
  static final FilteringTextInputFormatter _digitOnlyFormatter = FilteringTextInputFormatter.allow(_digitOnlyRegex);

  RandomNumberFormatter();

  @override
  String _formatPattern(String digits) {
    StringBuffer buffer = StringBuffer();

    for (int i = 0; i < digits.length; i++) {
      buffer.write(math.Random().nextInt(9));
    }
    return buffer.toString();
  }

  @override
  TextEditingValue _formatValue(TextEditingValue oldValue, TextEditingValue newValue) {
    return _digitOnlyFormatter.formatEditUpdate(oldValue, newValue);
  }

  @override
  bool _isUserInput(String s) {
    return _digitOnlyRegex.firstMatch(s) != null;
  }
}

///
/// An abstract class extends from [TextInputFormatter] and does numeric filter.
/// It has an abstract method `_format()` that lets its children override it to
/// format input displayed on [TextField]
///
abstract class NumberInputFormatter extends TextInputFormatter {
  TextEditingValue? _lastNewValue;

  @override
  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) {
    /// nothing changes, nothing to do
    if (newValue.text == _lastNewValue?.text) {
      return newValue;
    }
    _lastNewValue = newValue;

    /// remove all invalid characters
    newValue = _formatValue(oldValue, newValue);

    /// current selection
    int selectionIndex = newValue.selection.end;

    /// format original string, this step would add some separator
    /// characters to original string
    final newText = _formatPattern(newValue.text);

    /// count number of inserted character in new string
    int insertCount = 0;

    /// count number of original input character in new string
    int inputCount = 0;
    for (int i = 0; i < newText.length && inputCount < selectionIndex; i++) {
      final character = newText[i];
      if (_isUserInput(character)) {
        inputCount++;
      } else {
        insertCount++;
      }
    }

    /// adjust selection according to number of inserted characters staying before
    /// selection
    selectionIndex += insertCount;
    selectionIndex = math.min(selectionIndex, newText.length);

    /// if selection is right after an inserted character, it should be moved
    /// backward, this adjustment prevents an issue that user cannot delete
    /// characters when cursor stands right after inserted characters
    if (selectionIndex - 1 >= 0 && selectionIndex - 1 < newText.length && !_isUserInput(newText[selectionIndex - 1])) {
      selectionIndex--;
    }

    return newValue.copyWith(text: newText, selection: TextSelection.collapsed(offset: selectionIndex), composing: TextRange.empty);
  }

  /// check character from user input or being inserted by pattern formatter
  bool _isUserInput(String s);

  /// format user input with pattern formatter
  String _formatPattern(String digits);

  /// validate user input
  TextEditingValue _formatValue(TextEditingValue oldValue, TextEditingValue newValue);
}
```
</details>

If I swap in `CreditCardFormatter` for the `RandomNumberFormatter`, the assertions don't get thrown, so it seems like this is a reasonable solution.

justinmc replied:
I tried this out again and it seems to work for me on both iOS and Android with Gboard, where it correctly includes composing state in the undo history.  I also tried the example above and got the assertion.  Thanks!

fbcouch committed:
PR Updates

fbcouch committed:
Fix trailing spaces

fbcouch committed:
Where did that dill file even come from

fbcouch commented:
@justinmc @chunhtai I believe this is ready for another look. It looks like there's some infrastructure problem with that docs_test. Also, I tried to incorporate the changes from https://github.com/flutter/flutter/pull/108765 to make sure we don't have a regression there.

fbcouch committed:
Pass a duringTrigger bool to the shouldChangeUndoStack callback to allow opting out of text input formatter changes while undoing

fbcouch committed:
Merge branch 'main' into feature/undo-redo-gestures

chunhtai reviewed (approved):
left a reply with a suggestion, otherwise this LGTM

fbcouch committed:
Remove duringTrigger variable from the shouldChangeUndoStack signature, instead encoding that behavior in _push

fbcouch committed:
Adding assertion to ensure value is either the nextValue or lastValue after onTriggered call

justinmc commented:
@fbcouch Can you update this branch by merging in master?  GitHub is saying there is a conflict, though it doesn't say what the conflict is...

Otherwise I think we need to merge this!

jmagman commented:
Thanks for sticking with this for so long, @fbcouch!

justinmc commented:
I spoke a little too soon, we should make sure that [this comment thread](https://github.com/flutter/flutter/pull/98294#discussion_r1063838774) is resolved.  I just left a comment explaining the composing text behavior.

fbcouch committed:
Merge branch 'main' into feature/undo-redo-gestures

fbcouch committed:
Include composing changes in the undo history

fbcouch committed:
Merge branch 'main' into feature/undo-redo-gestures

fbcouch committed:
Fix failing tests

fbcouch committed:
Fix analyze issue

fbcouch commented:
@justinmc @chunhtai I merged main into this branch again, trying to preserve relevant changes to the old TextEditingHistory from https://github.com/flutter/flutter/pull/120889, https://github.com/flutter/flutter/pull/120062, and https://github.com/flutter/flutter/pull/119028

I think we should have everything wrapped up here (see: https://github.com/flutter/flutter/pull/98294/files#r1080613858 )

justinmc reviewed (approved):
Renewing my LGTM with the latest changes 👍 

Thanks for doing all the work to bring this back up to date.  I'll plan to merge it tomorrow morning if no one else objects.
